delfin/drivers/dell_emc/power_store/rest_handler.py

```diff
@@ -0,0 +1,958 @@
+        for appliance in appliances:
+            pool_id = appliance.get('id')
+            pool_name = appliance.get('name')
+            if not pool_id or not pool_name:
+                continue
+            data = {'entity': consts.PERFORMANCE_METRICS_BY_APPLIANCE,
+                    'entity_id': pool_id,
+                    'interval': consts.PERFORMANCE_METRICS_INTERVAL}
+            packaging_data = self.package_data(data, end_time, start_time)
+            self.set_metrics_data(pool_id, pool_name, packaging_data,
+                                  resource_metrics,
+                                  constants.ResourceType.STORAGE_POOL,
+                                  storage_id,
+                                  pool_metrics)
+        return pool_metrics
+
+    def get_volume_metrics(self, storage_id, resource_metrics, start_time,
+                           end_time):
+        volume_metrics = []
+        volumes = self.rest_call(self.REST_VOLUME_URL)
+        for volume in volumes:
+            volume_id = volume.get('id')
+            volume_name = volume.get('name')
+            if not volume_id or not volume_name:
+                continue
+            data = {'entity': consts.PERFORMANCE_METRICS_BY_VOLUME,
+                    'entity_id': volume_id,
+                    'interval': consts.PERFORMANCE_METRICS_INTERVAL}
+            packaging_data = self.package_data(data, end_time, start_time)
+            self.set_metrics_data(volume_id, volume_name, packaging_data,
+                                  resource_metrics,
+                                  constants.ResourceType.VOLUME, storage_id,
+                                  volume_metrics)
+        return volume_metrics
+
+    def get_controllers_metrics(self, storage_id, resource_metrics, start_time,
+                                end_time):
+        controllers_metrics = []
+        controller_d = self.get_node_hardware()
+        controllers = self.rest_call(self.REST_NODE_URL)
+        for controller in controllers:
+            controller_id = controller.get('id')
+            if not controller_id:
+                continue
+            hardware_id, hardware_name = self.get_resource(controller,
+                                                           controller_d)
+            if not hardware_id:
+                LOG.info('controllers performance: Unexpected data')
+                continue
+            data = {'entity': consts.PERFORMANCE_METRICS_BY_NODE,
+                    'entity_id': controller_id,
+                    'interval': consts.PERFORMANCE_METRICS_INTERVAL}
+            packaging_data = self.package_data(data, end_time, start_time)
+            self.set_metrics_data(hardware_id, hardware_name, packaging_data,
+                                  resource_metrics,
+                                  constants.ResourceType.CONTROLLER,
+                                  storage_id,
+                                  controllers_metrics)
```

Comment:
We should avoid changing the value of the parameter in the function, we should try to use like:
```
controller_metric_data = self.get_metrics_data(...)
controllers_metrics.extend(controller_metric_data)
```

Reply:
There is no way to modify it. This function is a public method and cannot be extracted without this parameter